[Static Analyzer CI] Add smart pointer analysis queue to EWS

briannafan opened the pull request:
#### 2d68279abbd5dc5a27a5f59521fbb1fbbbb83c7f
<pre>
[Static Analyzer CI] Add smart pointer analysis queue to EWS
<a href="https://bugs.webkit.org/show_bug.cgi?id=276625">https://bugs.webkit.org/show_bug.cgi?id=276625</a>
<a href="https://rdar.apple.com/131782284">rdar://131782284</a>

Reviewed by Aakash Jain.

* Tools/CISupport/ews-build/factories.py:
(SmartPointerStaticAnalyzerFactory.__init__): Added.

* Tools/CISupport/ews-build/steps.py:
(RevertAppliedChanges.__init__): Add exclude argument to ignore given patterns.
(RevertAppliedChanges.run):

(ScanBuildSmartPointer): Copied from build-webkit with following differences:
  - baseline arg added for building without PR changes applied
  - Call FindUnexpectedStaticAnalyzerResults with expectations = False if baseline build

(ParseStaticAnalyzerResults): Copied from build-webkit. Added baseline arg.

(FindUnexpectedStaticAnalyzerResults): Copied from build-webkit. Added expectations arg.
(FindUnexpectedStaticAnalyzerResults.run):
  - Run script with --check-expectations arg if expectations = True.
    If there are failures, rebuild without changes.
  - If expectations = False, compare PR results with baseline results.
(FindUnexpectedStaticAnalyzerResults.createResultMessage):

(DisplaySmartPointerResults): Copied from build-webkit. Slight differences in wording.
(DisplaySmartPointerResults.getFilesPerProject):
(DisplaySmartPointerResults.getResultSummary): Comment and block the PR if there are failures.

(PrintClangVersion): Copied from build-webkit with minor changes.
(ArchiveStaticAnalyzerResults): Ditto.
(UploadStaticAnalyzerResults): Ditto.
(ExtractStaticAnalyzerTestResults): Ditto.

Tools/CISupport/ews-build/events.py: Added steps to STEPS_TO_REPORT.
* Tools/CISupport/ews-build/steps_unittest.py:
* Tools/CISupport/ews-build/config.json:
* Tools/CISupport/ews-build/loadConfig.py:

Canonical link: <a href="https://commits.webkit.org/284319@main">https://commits.webkit.org/284319@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c6ae9b2d0bb03796b1938545ae873073a2201d9c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69053 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48453 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21725 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/73134 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/20211 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/71170 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56254 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20060 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/73134 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/20211 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72119 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/44232 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/59617 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/73134 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/40898 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/17047 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/18585 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/62847 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/17392 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/74843 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/87/builds/13035 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/16637 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/74843 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/68731 "Passed tests") | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/86/builds/13073 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/59700 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/74843 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/4122 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10547 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/44257 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/45330 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/46526 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/45072 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -6874,3 +6902,311 @@ def doStepIf(self, step):
+    haltOnFailure = True
+    resultMsg = ''
+
+    def __init__(self, baseline=False, **kwargs):
```
please also set `logEnviron=False` to all these steps.

aj062 reviewed (commented):
Before landing and deploying, please make sure that the queue is consistently green and the failures are expected (e.g.: Issue with the PR being tested). Also, it should be clear to engineers what the failures indicate so that they are actionable.

One difference in ews from build.webkit.org is that ews compares the issues with the existing ones and finds the new issues and fails the build only for new issues. I haven't checked in too much detail, but please make sure that this behavior is working properly in this queue.

Separately, I checked a random recent build https://ews-build.webkit-dev.org/#/builders/184/builds/1581, and the '[View failures](https://ews-build-dev.s3-us-west-2.amazonaws.com/macOS-Sonoma-Smart-Pointer-Build-EWS/999ee475-1581/scan-build-output/unexpected-results.html)'  link is giving access denied.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -5923,6 +5928,29 @@ def getResultSummary(self):
         return {'step': configuration}
 
 
+class PrintClangVersion(shell.ShellCommandNewStyle):
```
Please set `logEnviron=False` here as well

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -6874,3 +6902,333 @@ def doStepIf(self, step):
+    bugs = 0
+
+    def __init__(self, baseline=False, **kwargs):
+        kwargs['timeout'] = 2 * 60 * 60
```
let's maintain consistency with this file, and specify timeout as parameter in `super().__init__()` instead of adding it to kwargs and then passing kwargs to that method.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -5923,6 +5928,29 @@ def getResultSummary(self):
+        log_text = self.log_observer.getStdout()
+        match = re.search('(.*clang version.+) (\\(.+?\\))', log_text)
+        if match:
+            return {'step': match.group(0)}
```
It looks somewhat ugly/non-readable to print the url, that too non-working url. Better to:
1)  construct a working url (e.g: modify `https://github.com/rniwa/llvm-project.git 3e13e8a92d8fa0037483352faa13d929ba83a5a6` => https://github.com/rniwa/llvm-project/commit/3e13e8a92d8fa0037483352faa13d929ba83a5a6)
2) add the url to the step as a clickable url (using `self.addURL()`)

Can do the same thing for this code in build.webkit.org as well. Although all this change can be a separate PR as well if you prefer that way.

briannafan replied:
This is a good idea. I created a bug to track this for a separate PR.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -6874,3 +6902,333 @@ def doStepIf(self, step):
+        status = ''
+        if self.analyzeFailed:
+            status += 'ANALYZE FAILED: '
+        status += self.name
```
Probably can skip the name of the step in the summary. Duplicating the name in the UI doesn't look very polished.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -6874,3 +6902,333 @@ def doStepIf(self, step):
+    def getResultSummary(self):
+        status = ''
+        if self.resultMsg:
+            status += f' Issues by project: {self.resultMsg}'
```
Can rename `Issues by project:` to `Issues:` to make the summary compact/focused.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -6874,3 +6902,333 @@ def doStepIf(self, step):
+        is_log += yield self.getFilesPerProject(unexpected_results_data, 'failures')
+        if not is_log and num_issues:
+            pluralSuffix = 's' if num_issues > 1 else ''
+            yield self._addToLog('stdio', f'Found {num_issues} new issue{pluralSuffix} in already failing files')
```
Can make this summary `Ignoring 1 new issue in already failing files` more consistent with rest of EWS (assuming that is accurate for this queue), something like: `Ignored 1 pre-existing issue`. So, that it's obvious to engineers who are used to that message.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -6874,3 +6902,333 @@ def doStepIf(self, step):
+    def getResultSummary(self):
+        status = ''
+        if self.analyzeFailed:
+            status += 'ANALYZE FAILED: '
```
Doesn't look polished to have this text in caps. For engineers what are they expected to interpret it as? Can we just drop this text and print the real error?

briannafan replied:
Yeah I meant to change this. It should be similar to `compile-webkit` to indicate a build failure. I changed it to `Failed to build and analyze WebKit`

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -6874,3 +6902,333 @@ def doStepIf(self, step):
+
+    def resultDirectoryURL(self):
+        path = self.resultDirectory.replace('public_html/results/', '') + '/'
+        return '{}{}'.format(S3_RESULTS_URL, path)
```
can use f-string here.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -6874,3 +6902,333 @@ def doStepIf(self, step):
+        num_issues = self.getProperty('unexpected_new_issues', 0)
+        results_link = self.resultDirectoryURL() + SCAN_BUILD_OUTPUT_DIR + "/unexpected-results.html"
+        build_link = f'{self.master.config.buildbotURL}#/builders/{self.build._builderid}/builds/{self.build.number}'
+        formatted_build_link = '[#{}]({})'.format(self.getProperty("buildnumber", ""), build_link)
```
can use f-string here

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -6874,3 +6902,333 @@ def doStepIf(self, step):
+        return {u'step': status}
+
+
+class DisplaySmartPointerResults(buildstep.BuildStep, AddToLogMixin):
```
would be nice to have unit-test as well for this step (and making sure all the steps being added have unit-tests).

briannafan replied:
Added!

aj062 commented on Tools/CISupport/ews-build/config.json:
```diff
@@ -172,6 +172,7 @@
     { "name": "ews228", "platform": "visionos-simulator-1" },
     { "name": "ews229", "platform": "visionos-simulator-1" },
     { "name": "ews230", "platform": "visionos-simulator-1" },
+    { "name": "ews237", "platform": "*" },
```
can add this bot separately as part of config changes, since that's kind of unrelated to the core logic.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -6874,3 +6902,333 @@ def doStepIf(self, step):
+        unexpected_results = self.getProperty('unexpected_failing_files', 0) or self.getProperty('unexpected_new_issues', 0) or self.getProperty('unexpected_passing_files', 0)
+        if self.expectations and unexpected_results:
+            # If there are unexpected results, rebuild without changes to verify causation
+            self.build.addStepsAfterCurrentStep([RevertAppliedChanges(exclude=['new*', 'scan-build-output*']), ScanBuildSmartPointer(baseline=True)])
```
why do we exclude certain things from RevertAppliedChanges?

briannafan replied:
We want to compare the results from the first scan-build run with the results from the second one after changes are reverted. These paths contain those results.

aj062 replied:
ok. Also, `scan-build-smart-pointer_1` (and similar steps if applicable) should be renamed to `scan-build-smart-pointer-without-change` (just like we name `run-layout-tests-without-change`, `run-api-tests-without-change`, `compile-webkit-without-change` etc. )

aj062 reviewed (commented):
In case of failure, the summary looks little confusing/non-actionable. For eg.: in https://ews-build.webkit-dev.org/#/builders/184/builds/1639 it says: `2 files with smart pointer errors 2 files fixed`. What does `2 files fixed` means? The summary doesn't say `45 new issues`. Also, for the step `display-smart-pointer-results` what's the meaning of `unexpected-passes` and `unexpected-failures`, these files are not even listed in the scan-build results report (https://ews-build-dev.s3-us-west-2.amazonaws.com/macOS-Sonoma-Smart-Pointer-Build-EWS/988497e8-1639/scan-build-output/StaticAnalyzerRegressions/WebCore/index.html). How can engineers know the details of the issues?

Separately, Is the developer expected to fix all those 45 issues, 2 files with errors and 2 files fixed, before they are allowed to land their PR? If so, might be a good idea to check with some PR authors for the PRs on which such failures are found, to make sure that the engineers understand how to read these results and act on them.

aj062 commented on Tools/CISupport/ews-build/factories.py:
```diff
@@ -97,6 +97,27 @@ def __init__(self, platform, configuration=None, architectures=None, triggers=No
         self.addStep(ApplyWatchList())
 
 
+class SmartPointerStaticAnalyzerFactory(factory.BuildFactory):
```
Please add unit-test for this Factory (in `factories_unittest.py`)

briannafan replied:
I will add this when I make the `config.json` changes to add a builder. If I add the unit-test for the factory without specifying a builder, it will error with `AssertionError: False is not true : Builder Smart-Pointer-Static-Analyzer doesn't exist, but has unnecessary expected steps`

aj062 commented on Tools/CISupport/ews-build/factories.py:
```diff
@@ -97,6 +97,27 @@ def __init__(self, platform, configuration=None, architectures=None, triggers=No
+        self.addStep(ConfigureBuild(platform=platform, configuration=configuration, architectures=architectures, buildOnly=buildOnly, triggers=triggers, triggered_by=triggered_by, remotes=remotes, additionalArguments=additionalArguments))
+        if checkRelevance:
+            self.addStep(CheckChangeRelevance())
+        self.addStep(ValidateChange(branches=self.branches))
```
Do we need to pass branches (and initialize branches = None) at all, as self.branches is None and the default value of branches is also None.

briannafan replied:
Good point. This was an artifact of following the `Factory` class initialization

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -2776,13 +2777,17 @@ class RevertAppliedChanges(steps.ShellSequence):
     haltOnFailure = True
 
-    def __init__(self, **kwargs):
+    def __init__(self, exclude=None, **kwargs):
```
Nit: Can consider adding unit-test in `TestRevertAppliedChanges` for exclude. Not a blocker though.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -5923,6 +5928,32 @@ def getResultSummary(self):
+    warnOnFailure = False
+
+    def __init__(self, **kwargs):
+        super().__init__(logEnviron=False, **kwargs)
```
Nit: can also add a timeout here, e.g.: `timeout=60`

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -5923,6 +5928,32 @@ def getResultSummary(self):
 
 
+class PrintClangVersion(shell.ShellCommandNewStyle):
+    name = 'print-clang-version'
```
Can consider adding unit-test for this class `TestPrintClangVersion`.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -5923,6 +5928,32 @@ def getResultSummary(self):
+    def run(self):
+        self.log_observer = logobserver.BufferLogObserver()
+        self.addLogObserver('stdio', self.log_observer)
+        self.command = ['../llvm-project/build/bin/clang', '--version']
```
Nit: Can move this above as a class variable to be more consistent with rest of the code.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -5923,6 +5928,32 @@ def getResultSummary(self):
+        if self.results != SUCCESS:
+            return {'step': 'Failed to print clang version'}
+        log_text = self.log_observer.getStdout()
+        match = re.search('(.*clang version.+) (\\(.+?\\))', log_text)
```
Nit: Can separate out the regular expression and define it as a class variable (e.g.: CLANG_VERSION_RE)

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -5923,6 +5928,32 @@ def getResultSummary(self):
+            return {'step': 'Failed to print clang version'}
+        log_text = self.log_observer.getStdout()
+        match = re.search('(.*clang version.+) (\\(.+?\\))', log_text)
+        if match:
```
Nit: Can consider adding a return statement to handle the case when the step is successful but the RE doesn't match.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -6874,3 +6905,330 @@ def doStepIf(self, step):
+            status += 'Failed to build and analyze WebKit'
+        if self.results == SUCCESS:
+            status += f'Found {self.bugs} issues'
+        return {u'step': status}
```
Nit: Do we need explicit `u` as this is Python3.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -6874,3 +6905,330 @@ def doStepIf(self, step):
+    description = ['parsing static analyzer results']
+    descriptionDone = ['parsed static analyzer results']
+    haltOnFailure = True
+    resultMsg = ''
```
Nit: `resultMsg` => `result_message`

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -6874,3 +6905,330 @@ def doStepIf(self, step):
+        if self.resultMsg:
+            status += f' Issues: {self.resultMsg}'
+        if self.results != SUCCESS:
+            status += f'{self.name} ({Results[self.results]})'
```
Nit: Do we need to include` {self.name}` in the summary text here.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -6874,3 +6905,330 @@ def doStepIf(self, step):
+    name = 'find-unexpected-static-analyzer-results'
+    description = ['finding unexpected static analyzer results']
+    descriptionDone = ['found unexpected static analyzer results']
+    resultMsg = ''
```
Ditto

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -6874,3 +6905,330 @@ def doStepIf(self, step):
+
+    def resultDirectoryURL(self):
+        path = self.resultDirectory.replace('public_html/results/', '') + '/'
+        return f'{S3_RESULTS_URL}{path}'
```
Nit: can combine these two lines in one.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -6874,3 +6905,330 @@ def doStepIf(self, step):
+            comment = f"Smart Pointer Build {formatted_build_link}: Found [{num_issues} new failure{pluralSuffix}]({results_link}), blocking PR #{self.getProperty('github.number')}.\n"
+            comment += 'Please address these issues before landing. See [WebKit Guidelines for Safer C++ Programming](https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines).\n(cc @rniwa)'
+            self.setProperty('comment_text', comment)
+            self.build.addStepsAfterCurrentStep([LeaveComment(), SetCommitQueueMinusFlagOnPatch(), BlockPullRequest()])
```
Are you supporting patch-workflow with this queue? If so, please test it with patch workflow as well. If not, that's fine and you can remove `SetCommitQueueMinusFlagOnPatch()`.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -6874,3 +6905,330 @@ def doStepIf(self, step):
+            comment += 'Please address these issues before landing. See [WebKit Guidelines for Safer C++ Programming](https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines).\n(cc @rniwa)'
+            self.setProperty('comment_text', comment)
+            self.build.addStepsAfterCurrentStep([LeaveComment(), SetCommitQueueMinusFlagOnPatch(), BlockPullRequest()])
+            results_summary = f'Found {num_issues} new failure{pluralSuffix} in {failing_files}'
```
What if there are thousands of issues found, this summary would be extremely long containing thousands of file-names, and the UI would be extremely messy. Same thing for `passing_files`.

To handle that, while generating `failing_files` above, truncate it to a set limit (e.g.: 50). e.g.: `failing_files = (", ").join(self.getProperty('failures', [])[:50])`

e.g.: https://github.com/WebKit/WebKit/blob/main/Tools/CISupport/ews-build/steps.py#L3093

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -6874,3 +6905,330 @@ def doStepIf(self, step):
+        return {u'step': results_summary}
+
+
+class ArchiveStaticAnalyzerResults(shell.ShellCommandNewStyle):
```
Please add unit-test TestArchiveStaticAnalyzerResults

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -6874,3 +6905,330 @@ def doStepIf(self, step):
+
+class UploadStaticAnalyzerResults(UploadTestResults):
+    name = "upload-static-analyzer-results"
+    workersrc = "/tmp/static-analysis.zip"
```
Better to store this path as a global variable and access that in multiple classes. Currently the risk is that someone might change it (in future) in one place and forget to update in another.

aj062 reviewed (commented):
Overall looks fine. Few comments inline.

Also, I had a quick look at one of the sample run: https://ews-build.webkit-dev.org/#/builders/184/builds/1743 which says: `Found 47 new failures in ...`

The report in https://ews-build-dev.s3-us-west-2.amazonaws.com/macOS-Sonoma-Smart-Pointer-Build-EWS/1a320cee-1743/scan-build-output/StaticAnalyzerRegressions/WebCore/index.html lists various issues in `background-fetch/BackgroundFetch.cpp` file, e.g.: line 381	, 374, 407, 319. But none of those lines are being modified by the commit being tested which is https://github.com/WebKit/WebKit/commit/1a320ceee87f51a8a37192b014ffc5bd4f895009


This indicate that the errors for which PR is being blamed might be pre-existing and that PR shouldn't be blamed for those. These might not be easily actionable by the PR author. Please check, although that doesn't directly block this PR, it would probably block brining this queue to production.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -75,6 +75,9 @@
 MSG_FOR_EXCESSIVE_LOGS = f'Stopped due to excessive logging, limit: {THRESHOLD_FOR_EXCESSIVE_LOGS_DEFAULT}'
+SCAN_BUILD_OUTPUT_DIR = 'scan-build-output'
+LLVM_DIR = 'llvm-project'
+STATIC_ANALYSIS_ARCHIVE = '/tmp/static-analysis.zip'
```
Nit: STATIC_ANALYSIS_ARCHIVE => STATIC_ANALYSIS_ARCHIVE_PATH

briannafan committed:
[Static Analyzer CI] Add smart pointer analysis queue to EWS
https://bugs.webkit.org/show_bug.cgi?id=276625
rdar://131782284

Reviewed by Aakash Jain.

* Tools/CISupport/ews-build/factories.py:
(SmartPointerStaticAnalyzerFactory.__init__): Added.

* Tools/CISupport/ews-build/steps.py:
(RevertAppliedChanges.__init__): Add exclude argument to ignore given patterns.
(RevertAppliedChanges.run):

(ScanBuildSmartPointer): Copied from build-webkit with following differences:
  - baseline arg added for building without PR changes applied
  - Call FindUnexpectedStaticAnalyzerResults with expectations = False if baseline build

(ParseStaticAnalyzerResults): Copied from build-webkit. Added baseline arg.

(FindUnexpectedStaticAnalyzerResults): Copied from build-webkit. Added expectations arg.
(FindUnexpectedStaticAnalyzerResults.run):
  - Run script with --check-expectations arg if expectations = True.
    If there are failures, rebuild without changes.
  - If expectations = False, compare PR results with baseline results.
(FindUnexpectedStaticAnalyzerResults.createResultMessage):

(DisplaySmartPointerResults): Copied from build-webkit. Slight differences in wording.
(DisplaySmartPointerResults.getFilesPerProject):
(DisplaySmartPointerResults.getResultSummary): Comment and block the PR if there are failures.

(PrintClangVersion): Copied from build-webkit with minor changes.
(ArchiveStaticAnalyzerResults): Ditto.
(UploadStaticAnalyzerResults): Ditto.
(ExtractStaticAnalyzerTestResults): Ditto.

Tools/CISupport/ews-build/events.py: Added steps to STEPS_TO_REPORT.
* Tools/CISupport/ews-build/steps_unittest.py:
* Tools/CISupport/ews-build/config.json:
* Tools/CISupport/ews-build/loadConfig.py:

Canonical link: https://commits.webkit.org/284319@main
